### PR TITLE
Do not reject disabled or deleted versions/files during a block

### LINKED ISSUE #16455: Add-on version is rejected again when an existing block is edited
### Describe the problem and steps to reproduce it:
1. With an admin account, submit a new add-on block
2. Check the review page of the blocked add-on - notice the Block information and the Reject message recorded in the version history
3. Edit the add-on block (i.e. update the Reason or change the Version range)
4. Check the add-on review page again

### What happened?
The version history shows a new Reject message below the Block edit information

### What did you expect to happen?
The Rejection message should appear only once, when the version block was first created. 

### Anything else we should know?
When a block is edited to include additional versions of the same add-on, we should still show a Reject message next to each additional version covered by the block range.

https://reviewers.addons.allizom.org/en-US/reviewers/review/devhub-release-08-11
image

### LINKED ISSUE #17035: Blocking a deleted version should not cause a rejection action
### Describe the problem and steps to reproduce it:

I noticed on prod that a version that had already been deleted received a rejection action when it got blocked.

Anticipated STR:
1. Submit a version
2. Delete it from the devhub
3. Block the deleted version.

### What happened?

On the reviewer tools page that a _Rejected_ action was recorded as part of the block.

### What did you expect to happen?

No _Rejected_ action since the file has already been deleted.

### Anything else we should know?
(Please include a link to the page, screenshots and any relevant files.)

@ioanarusiczki or @AlexandraMoga wouuld you mind checking if you can confirm this on -dev using the STR above, please?

### PR DESCRIPTION
**Pull request by wagnerand**
Fix #16455.
Fix #17035.

### CODE REVIEW

File: src/olympia/blocklist/tests/test_admin.py [hunk 1]
```diff
@@ -304,6 +313,14 @@ def test_add_single(self):
             f'The blocklist submission {FANCY_QUOTE_OPEN}No Sign-off: guid@; '
             f'dfd; some reason{FANCY_QUOTE_CLOSE} was added successfully.'
         ]
```
Comment (eviljeff) [inline src/olympia/blocklist/tests/test_admin.py:315]: This test is.... hard to follow, and I wrote it...  🐼 
So, things to change:
- lets not repeatedly assign `log` to a different thing.  Where we don't need to reference the log entry again just don't assign it a var.  Otherwise have sensible naming.
- assert that `len(logs) == 5` - and if it isn't, assert what the extras log entries are (so we know there aren't additional block or reject actions)
- if we're asserting that there's only 1 log for the disabled version we kinda need to assert there are the expected log entries for the other versions.  

I think the following is a better way of checking the entries as as expected (note, not tested, I'm just writing it directly in here!):

```
        logs = ActivityLog.objects.for_addons(addon)
        assert len(logs) == 5
        assert logs[0].action == amo.LOG.CHANGE_STATUS.id
        assert logs[1].action == amo.LOG.CHANGE_STATUS.id
        reject_log1 = logs[2]
        assert reject_log1.action == amo.LOG.REJECT_VERSION.id
        reject_log2 = logs[3]
        assert reject_log2.action == amo.LOG.REJECT_VERSION.id
        block_log = logs[4]
        assert block_log.action == amo.LOG.BLOCKLIST_BLOCK_ADDED.id
        assert block_log.arguments == [addon, addon.guid, block]
        assert block_log.details['min_version'] == '0'
        assert block_log.details['max_version'] == addon.current_version.version
        assert block_log.details['reason'] == 'some reason'
        assert block_log == (
            ActivityLog.objects.for_block(block).filter(action=block_log.action).first()
        )
        assert block_log == (
            ActivityLog.objects.for_guidblock('guid@').filter(action=block_log.action).first()
        )
        assert [reject_log_1, block_log] == list(ActivityLog.objects.for_versions(first_version))
        assert [reject_log_2, block_log] == list(ActivityLog.objects.for_versions(second_version))
        assert [block_log] == list(ActivityLog.objects.for_versions(deleted_addon_version))
        assert not ActivityLog.objects.for_versions(pending_version).exists()
```

File: src/olympia/blocklist/tests/test_admin.py [hunk 2]
```diff
@@ -304,6 +313,14 @@ def test_add_single(self):
...
+        # The disabled and deleted versions should only have the block activity,
+        # not a reject activity
+        log_disabled_version = ActivityLog.objects.for_versions(disabled_version)
+        assert len(log_disabled_version) == 1
+        assert log == log_disabled_version.last()
+        log_deleted_version = ActivityLog.objects.for_versions(deleted_version)
+        assert len(log_deleted_version) == 1
+        assert log == log_deleted_version.last()
```
Comment (eviljeff) [inline src/olympia/blocklist/tests/test_admin.py:323]: could you change the tests here to be in a similar style as the previous comment?  (functionally they'll be the same, but it'd be more consistent)

**Review by eviljeff — COMMENTED** [review]
some test rewrites for glorious future.

File: src/olympia/blocklist/utils.py [hunk 3]
```diff
@@ -119,7 +119,10 @@ def disable_addon_for_block(block):
                 ver
                 for ver in block.addon_versions
                 # We don't need to reject versions from older deleted instances
-                if ver.addon == block.addon and block.is_version_blocked(ver.version)
+                # and already disabled files
+                if ver.addon == block.addon
+                and block.is_version_blocked(ver.version)
+                and ver.file.status != amo.STATUS_DISABLED
```
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/blocklist/utils.py:125]: Is that even possible?

[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/blocklist/utils.py:125]: Hm, if I add ´and ver.deleted = False` in there, we wouldn't reject those, but also not update the wrong status. That does not seem desirable?

File: src/olympia/blocklist/tests/test_admin.py (hunk 1 above)
Comment (eviljeff) [reply src/olympia/blocklist/tests/test_admin.py:315]: > Wouldn't it make more sense to use `ActivityLog.objects.for_versions()`, and check the activity logs for each version?

`ActivityLog.objects.for_versions()` requires a `versions` arg though.  If you mean something like `ActivityLog.objects.for_versions([first_version, second_version, deleted_addon_version, disabled_version, deleted_version)` it seems like it'd be more difficult to sort out the response to know which activity log corresponded which version?

File: src/olympia/blocklist/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/blocklist/utils.py:125]: There's no database constraint enforcing that deleted versions have disabled files 😬 
It's an edge case though (all versions that are deleted _now_ have the file status set to disabled), so it could be ignored if you're okay with that?  I.e. just assume all deleted versions also have the file set to disabled.  (what your code is assuming)

I'd lean away from updating "wrong" statuses.

File: src/olympia/blocklist/tests/test_admin.py [hunk 4]
```diff
@@ -233,6 +233,15 @@ def test_add_single(self):
...
         )
         first_version = addon.current_version
+        disabled_version = version_factory(
+            addon=addon, version='2.5', file_kw={'status': amo.STATUS_DISABLED}
+        )
+        deleted_version = version_factory(
+            addon=addon,
+            version='2.5.1',
+            deleted=True,
+            file_kw={'status': amo.STATUS_DISABLED},
```
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/blocklist/tests/test_admin.py:243]: this depends on https://github.com/mozilla/addons-server/pull/19492/files#r926577723 - if we're okay with ignoring the edge case of a non-disabled file status deleted version then this is fine as is.

File: src/olympia/blocklist/tests/test_admin.py (hunk 1 above)
Comment (wagnerand) [reply src/olympia/blocklist/tests/test_admin.py:319]: Yes, I just left out the argument for brevity. Since this is an isolated test and we define which versions we create and what versions apply to the block, it should be easy to write that down? I feel that would make reading and understanding the test a lot clearer.

Rather than getting the activity logs for all versions at once, we'd get the logs for each version separately and compare the length and individual items to what we'd expect based on what should have happened to the version during the block (block activity vs block activity+reject activity).

File: src/olympia/blocklist/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wagnerand) [reply src/olympia/blocklist/utils.py:125]: On prod there are currently 11 deleted versions with files that are approved. I checked one and it was because the version was deleted in between the auto-approval cron determining the files to be approved and the actual approval, which we fixed in #19136.

I'd leave this as is and fix the 11 versions on prod manually, if that works for you?

[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/blocklist/utils.py:125]: wfm.

**Review by eviljeff — APPROVED** [review]
